### ISSUE TITLE
GUI Interface Request

### ISSUE THREAD

**Issue by scamille** [scamille/simc_issue_test#39]
_From [DigitalS...@gmail.com](https://code.google.com/u/111723506478478800532/) on March 26, 2009 23:04:41_

Could ya possibly get a GUI for the Windows Version?

_Original issue: http://code.google.com/p/simulationcraft/issues/detail?id=40_

**Comment by scamille** [comment]
_From [natehieter@gmail.com](https://code.google.com/u/natehieter@gmail.com/) on March 27, 2009 18:38:05_

It is likely the codebase will be ported to Qt sometime in the near future.  At that
point we will consider a GUI..... but it may not happen until we have complete
support for all classes.

**Summary:** GUI Interface Request  
  **Status:** Accepted

**Comment by scamille** [comment]
_From [natehieter@gmail.com](https://code.google.com/u/natehieter@gmail.com/) on March 27, 2009 18:53:15_

**Owner:** natehieter  
  **Labels:** -Type-Defect Type-Enhancement

**Comment by scamille** [comment]
_From [triplepoint](https://code.google.com/u/triplepoint/) on April 10, 2009 21:02:58_

Would we be interested in a simple php frontend?  

It would just be a matter of taking form input from an HTML/javascript UI, generating
a config string on the fly in PHP to run simcraft, and then showing the html output
of simcraft to the user.

I only found this project yesterday; is there some plan for a UI, besides what nate
said above about QT?  Is there a good reason not to go the web-app route?

**Comment by scamille** [comment]
_From [natehieter@gmail.com](https://code.google.com/u/natehieter@gmail.com/) on April 11, 2009 00:47:07_

The sim is certainly fast enough for single-player DPS calcs on the fly....... 
However, a setup that could generate scale factors under a heavy user load might be
tricky.  If you're talking about running on the local machine, then the issue of user
load is gone.

I'm certainly interested in a PHP setup.  GUI stuff is very far from my area of
expertise, so I've been leaving it til last.  Anything that I envision ends up
looking alot like the "less is more" category......

**Comment by scamille** [comment]
_From [triplepoint](https://code.google.com/u/triplepoint/) on April 11, 2009 11:27:35_

Ha, I know what you mean regarding UI. I'm a professional web-app developer, and I'm
still not sure I've mastered it.

I wouldn't want to get into the business of hosting a web service, but I think it's
not too crazy to ask people to install the WAMP or LAMP stack on their local machine,
and hit the UI locally.

I'll take a look at it and see how feasible it is.

**Comment by scamille** [comment]
_From [natehieter@gmail.com](https://code.google.com/u/natehieter@gmail.com/) on April 11, 2009 11:53:45_

Excellent!  I'm very curious to see what is possible in this arena.

**Comment by scamille** [comment]
_From [triplepoint](https://code.google.com/u/triplepoint/) on April 20, 2009 16:50:05_

I've got a php-based UI working that covers all the options that the command-line
tool can accept (I think).  It's a little clunky, but it's functional.

  I've got it hosted off my desktop at home right now, so I'm reluctant to post the
link here.  Is there a way I can get a link to you?

  Also, which would be preferable: posting my project as a sub-directory of this
project, or starting a separate code.google sister project?

I'll attach a screenshot here, so everyone can at least see the current status.

**Comment by scamille** [comment]
_From [triplepoint](https://code.google.com/u/triplepoint/) on April 20, 2009 16:54:21_

whoops. screenshot i had posted had confidential info.  Here's the revised version.

**Attachment:** [Screenshot-SimulationCraft - Mozilla Firefox-3.png](http://code.google.com/p/simulationcraft/issues/detail?id=40#hc8)

**Comment by scamille** [comment]
_From [triplepoint](https://code.google.com/u/triplepoint/) on April 20, 2009 16:56:37_

I should point out, this isn't the final version.  This is more of a milestone
release, just to get the ball rolling.

**Comment by scamille** [comment]
_From [natehieter@gmail.com](https://code.google.com/u/natehieter@gmail.com/) on April 20, 2009 18:21:34_

I can be reached pretty much 24hrs a day at natehieter@gmail.com

The SS looks interesting: I didn't expect you to go after multi-player sim, too. 
This is much more ambitious!  

I imagine there are two distinct use-models:
(1) Multi-player: focusing on easy/smart defaults, enabling fast raid composition
with high-level choices on specs, gear, and actions.
(2) Single-player: focusing on Armory interaction to get specific talents and gear,
enabling creative exploration of action priorities/conditionals, generating scale
factors and gear up-grades

As for hosting, that is up to you: There are benefits to both internal and "sister"
setups.  If it gets checked in here we guarantee dev overlap and issue communication.
 I think the devs here will vouch for my easy-going attitude: I'm no dictator! 
But....  I feel an obligation to review/guide any contributions made here.

As for future direction, I still would like to port to QT, but the reality of that
happening any time soon is hard to imagine.  Additionally, I'm fascinated by the
possibilities of a server/client setup that your php implementation allows.

Whichever way you choose, I've set you up for commit access here and have invited you
to the simulationcraft googlegroup mailing list.

**Comment by scamille** [comment]
_From [Twigele](https://code.google.com/u/Twigele/) on April 21, 2009 07:52:33_

I have to say that screenshot looks awesome, wow. I'm a real gimp at doing GUI things
(as creative as a slice of bread), nice job, I'd really like to see it in action.

**Comment by scamille** [comment]
_From [natehieter@gmail.com](https://code.google.com/u/natehieter@gmail.com/) on April 21, 2009 08:08:37_

triplepoint: If you find there are setup aspects that make your job
difficult/awkward, let me know.  I'm more than willing to make changes on the
engine-side to enable cleaner implementations on the gui-side.

Things that jump to mind: 
It would probably be a good idea to separate out the OOC actions from the action list.  
Pulling out flask/food into parameters might help, too.

**Comment by scamille** [comment]
_From [zmnspencer@gmail.com](https://code.google.com/u/zmnspencer@gmail.com/) on July 14, 2009 12:20:22_

Is this still being worked on?

I looked at the version on svn and I can see some issues.

I would be willing to work on this myself if desired.

**Comment by scamille** [comment]
_From [natehieter@gmail.com](https://code.google.com/u/natehieter@gmail.com/) on July 14, 2009 16:34:23_

triplepoint has been moving things along as he has time.  But I think the biggest
slowdown has been virtually no guidance from me...... because I simply have been
unable to envision a workable gui.

The advent of the wowhead site support for player profiles has changed all that.... 
It enables us to defer much of the un-fun complication of GUIs by reusing their site.
 Of course, it means that the GUI could not run disconnected from the internet, but
that is a fair compromise.

Consider a very simple window with tab-frames and a text "command line" at the bottom.

Tab Frame 1: Various (limited) top-level parameters
Tab Frame 2: Armory web-page
Tab Frame 3: Wowhead Profiler web-page
Tab Frame 4: Multi-line text editor
Tab Frame 5: HTML results

To the right of the command-line at the bottom is a simple "Go!" button.

If the active tab is the Armory web-page, then we download and run with whatever
character is being viewed.  If the active tab is the Wowhead profiler web-page, then
we download and run with that character ID.

While the Armory tab would be rather static, the wowhead tab would allow all sorts of
gear/gem/enchant/talent experimentation.  Just click "Save" on their site, and "Go!"
in our window.  This would download the profile, populate tab frame `#4` with the
config info, run the sim, and then populate tab frame `#5` with the results.

While this loses the one of the more interesting features of both the sim and the
current GUI effort, multi-player interaction, I think concentrating on single-player
ease-of-use and intuitive report generation is a fair trade-off.  And of course,
populating multi-player sims is a natural follow-on.

I'd love to be able to do the same with chardev, but I've yet to decode their rather
cryptic setup.

As for choosing a platform to implement this..... I don't think it matters.  Ideally,
we'd have several alternatives.  The current PHP/XML implementation is cool because
it could be setup server-side if desired.  However, a Java implementation via JNI or
a Qt app wrapper are also very interesting.

It doesn't matter because the HTML output generation would be common and the majority
of the input parameters would also be common via armory/wowhead/chardev.  This leaves
very light wrappers that, quite frankly, could be pursued in parallel depending upon
the level of interest: We're in this for fun, not to make a buck.

**Comment by scamille** [comment]
_From [natehieter@gmail.com](https://code.google.com/u/natehieter@gmail.com/) on July 14, 2009 16:45:14_

Long-story short: There is no real master-plan for SimulationCraft.  It is an organic
app that grows in fits and starts as people come and go.  You are all volunteers, so
I tend to pretty open about letting new folks stretch the app in new directions.  Of
course, I have a habit of re-re-re-writing my own and other contributions until I'm
happy, but I try not to throw out any "thou shalt"s.

If this loosey-goosey development environment hasn't scared you off, contact me
directly and I'll set you up with commit access.  If you want to make changes to the
current GUI setup I strongly encourage you to coordinate with triplepoint.  If you
want to explore another direction that's fine, too.  

In general, I prefer to give people a ton of freedom to work in order to ensure folks
have a chance to really explore their idea.  If it is solid, then I'll take a hard
look at their prototype and plan out a production re-write.

**Comment by scamille** [comment]
_From [zmnspencer@gmail.com](https://code.google.com/u/zmnspencer@gmail.com/) on July 14, 2009 17:53:05_

natehieter, I've have been approaching the whole GUI problem from a different, complemental, angle: in-
game, via an addon. While this method has clear drawbacks - all communication has to be by hand, it also 
allow real time data, and means that the users doesn't have to worry about what a player spec is or which gear 
set they have on on armory.

The ultimate goal here is being to evaluate gear for the whole raid, even if roughly. As such, my intermediate 
goal is to able to pick any arbitrary player in game, and with a few copy & pastes, be able to get -  personal - 
scaling factors, without the player having to tell me anything or me having to make any choices.

Given all this, my goal with the tickets/comments here is to make my life easier with some changes to the web 
UI and/or Simcraft. For example, being able to paste a parameter string (or any format of string really) in the 
web UI, hit "Go" and get a copyable result would remove all the temp file/command line messiness needed 
now.

In closing, if this is in active development, I'm not going to meddle. However, there are some 
improvement/bug fixes that I would like to see in the web UI. Where should I put them? Should I make a new 
ticket for each? Or should I just make comments in the ticket?

**Comment by scamille** [comment]
_From [natehieter@gmail.com](https://code.google.com/u/natehieter@gmail.com/) on July 14, 2009 18:12:27_

I think "active development" is perhaps too strong.  I think "periodic development"
is more descriptive.  All of us are working on this in our "spare" time and sometimes
there is less than others.  The GUI suffers from the simple fact that only one person
is contributing, so when he is otherwise occupied nothing happens.  By contrast, the
engine has several people pinch hitting across various modules.

If you are looking bug fixes and minor feature enhancements, then by all means feel
free to simply commit them.  I'll happily provide you with commit access.

**Comment by scamille** [comment]
_From [triplepoint](https://code.google.com/u/triplepoint/) on July 14, 2009 21:56:05_

zmnspencer, 

I have to apologize for the lack of recent commits.  I've been doing a bit of
overhauling of the config form's XML structure as derived from the C++ source, and
its not in a clean enough state to commit yet.

I'm certainly open to feature suggestions or code commits on the PHP/XSL code.  I've
done my best to keep all the various code elements clean and well documented so that
anyone with some experience should be able to pick it up.

I agree that the GUI as it stands is not fully functional, and there're some layout
issues left to handle.  The biggest problem with building a UI for this app is the
huge number of possible options to twiddle, and there's new ones appearing and old
ones disappearing every few builds.  Just getting the fields editable and usable took
some doing - the future development is certainly going to focus more on usability and
layout.

If what you had in mind was overhaul scale, I'd prefer to fork my current work and
develop independent forks for a while.  If you'd rather work together to hone the
existing work, then I'm certainly willing and able to go that route.  And of course,
if you want to start from scratch and implement something completely new (in another
language, or as a standalone app, say), there's certainly plenty of room for
alternatives.

Feel free to email me directly at my user name at gmail.com if you want to discuss
the particulars, or if you join the group there's plenty of room for discussion on
our forum.

And of course, anyone else that reads this is always welcome to contact me for
conversation on the GUI development.  I can't always promise immediate responses, but
I can usually reply within 24 hours or so.

**Comment by scamille** [comment]
_From [natehieter@gmail.com](https://code.google.com/u/natehieter@gmail.com/) on July 15, 2009 06:12:35_

I appreciate the effort, triplepoint.  SimulationCraft is not just a moving
target.....  It dodges and dives!

**Comment by scamille** [comment]
_From [zmnspencer@gmail.com](https://code.google.com/u/zmnspencer@gmail.com/) on July 25, 2009 16:22:32_

triplepoint, is what you committed last week the overhaul you were referring to?

In any case, I never got an answer for where to put issues/requests for the GUI so I will just put them here.

1) The server drop-down for "Add from Armory" is way too long and hard to use. Furthermore, as it now, 
typing to select a server fails in Safari (It always selects the last option regardless of what you type). It doesn't 
like the optgroup.

2) Servers with am apostrophe do not work with "Add From Armory". The server names needs to be html-
escaped in the input to simcraft.

3) Support for armory talent selection ( https://code.google.com/p/simulationcraft/issues/detail id=93&can=1)

4) Some way to import SimCraft by copy/paste into an input field

**Comment by scamille** [comment]
_From [triplepoint](https://code.google.com/u/triplepoint/) on July 25, 2009 18:23:07_

The structural XML changes from last week were intended to make it easier to
customize portions of the GUI.  The XML structure now reflects the structure of the
opt_doc comments scraped from the various C++ classes.

Yes, I agree the select box for server in Add from Armory is too large.  The best fix
I've seen on other sites is to have a second select for 'region'.  As for it failing
the type-find in Safari, that's of course a bug in Safari.  Still, breaking that
select up into two fields and removing the optgroup should fix that bug.

It's difficult to test the armory import for different servers, without knowing names
of characters on that server which fail import.  I'll focus on server names with
apostrophes in their names, and see if I can get a fix out.

The talent system in the GUI is definitely in need of a higher-level GUI.  This is
something I'm working on and will commit when it's even sort of usable.

There's functionality to import simcraft commands from an uploaded file; I hadn't
considered a text field for the same purpose.  If I can find a clean way to squeeze
it into the GUI I'll get this in soon, it should be an easy addition.

**Comment by scamille** [comment]
_From [zmnspencer@gmail.com](https://code.google.com/u/zmnspencer@gmail.com/) on July 25, 2009 19:19:59_

Sorry, I said apostrophe, but I was thinking space. A example of a character: http://www.wowarmory.com/character-sheet.xml?r=Laughing+Skull&n=Walikdu I attached a patch for the server issue.

Another minor request: when you import a character from Armory, can it be properly capitalized?

**Attachment:** [GUI_diff.patch](http://code.google.com/p/simulationcraft/issues/detail?id=40#hc22)

**Comment by scamille** [comment]
_From [negative...@gmail.com](https://code.google.com/u/117721594206628654566/) on September 09, 2009 15:11:16_

Just wanted to post that a guildy and i are already working a web-based php version for 
this, unless we get lazy it should be up within the next few weeks..

**Comment by scamille** [comment]
_From [natehieter@gmail.com](https://code.google.com/u/natehieter@gmail.com/) on September 09, 2009 15:26:17_

Sounds great!  When you say "this" are you referring to the existing php code, or a
parallel effort?

I have been meaning to work on a very simple QT skeleton front-end for months and
months..... but it keeps getting deferred in favor of model enhancement.  Thankfully,
I'm running out of low-hanging fruit so maybe something will happen soon.

**Comment by scamille** [comment]
_From [negative...@gmail.com](https://code.google.com/u/117721594206628654566/) on September 09, 2009 17:42:10_

Sorry, i am referring to a parallel effort. I just discovered simulationcraft last
night and have been messing with it ever sense. Very impressed so far.

Started by writing a simple perl script to run the calculations (for people who have
very little idea how command line and readme's work). Told a friend, he suggested the
php project and here we are. 

I'll continue posting updates as we go.

PS: Thanks for the warm welcome =)

**Comment by scamille** [comment]
_From [natehieter@gmail.com](https://code.google.com/u/natehieter@gmail.com/) on September 09, 2009 18:03:44_

Best of luck, feel free to send me email if you have any questions.  I have no grand
vision for the GUI so I'm perfectly fine with multiple threads.  After all, we do
this for fun......

If you reach the point where you feel it would be helpful to make some changes to the
core engine, let me know and I'll get you commit access.

**Comment by scamille** [comment]
_From [natehieter@gmail.com](https://code.google.com/u/natehieter@gmail.com/) on September 22, 2009 22:30:44_

I've got enough of the Qt gui framework in place that it is worth checking out.

Glaring holes:
(1) Globals page is ignored at the moment.
(2) No user-history management.  The "<" and ">" buttons currently only work on the
web view pages.  Eventually, these will iterate through past states for each of the
Globals, Simulate, and Overrides tabs.  Also, the up/down arrows at the cmd line will
iterate through past cmds.
(3) Save has not yet been implemented for log/results.
(4) Welcome page will have html-encoded help describing the import->simulate->browse
process.

**Comment by scamille** [comment]
_From [natehieter@gmail.com](https://code.google.com/u/natehieter@gmail.com/) on September 22, 2009 22:35:00_

Forgot to mention: For most tabs, the cmd line will act like the unix cmd line, but
without having to type "./simcraft" first.

For example, running ./simcraftqt will drop you at the Welcome tab.  Typing
"armory=us,llane,segv" followed by the ENTER-key will pass that string to the simulator. 

Note that the progress bar will appear to grow twice.  The first is for the import. 
The second is for the simulation.

**Comment by scamille** [comment]
_From [natehieter@gmail.com](https://code.google.com/u/natehieter@gmail.com/) on September 23, 2009 08:57:27_

Progress:

Regarding `#1`: Settings on the "Globals" page are now honored.
Regarding `#3`: Simplistic support for saving logs and html results.

Working on help next......

**Comment by scamille** [comment]
_From [negative...@gmail.com](https://code.google.com/u/117721594206628654566/) on September 28, 2009 19:59:09_

Just wanted to update about the php backend.

Unfortunately my partner for this project has backed out. 

So although it's not all that much work, i thought i'd share the bit of work i did already before i give up on it. 
Might help somebody.

Here is a perl script to make it a bit more convenient to run the program, essentially eliminates the use for 
command-line options and arguments. Figured it may help somebody out, or atleast save them some time. 
Appreciate the support offered and good luck with the rest of the project =) http://www.v3trae.net/documents/scriptcraft.txt

**Comment by scamille** [comment]
_From [natehieter@gmail.com](https://code.google.com/u/natehieter@gmail.com/) on October 21, 2009 11:11:57_

We have released a GUI wrapper now, so we no longer need a generic GUI request. 
GUI-related Issues can be addressed individually.

**Status:** Fixed
